### ISSUE TITLE
Stream Subscription stops receiving messages after timeout

### ISSUE THREAD

**Issue by rastrup** [EventStore/EventStore#3491]
**Describe the bug**
When using Stream subscriptions, the subscription times out after some time but subscriptionDropped is not being called and no new events are received.

**To Reproduce**
Set up a subscription using EventStoreClient.SubscribeToStreamAsync(...)
Let it run for a bit.
A timeout occurs.
The subscription stops receiving new messages but the subscriptionDropped Action is not invoked and no new events are processed.

**Expected behavior**
The subscriptionDropped Action would be invoked allowing us to restart the subscription or it would recover itself.

**Actual behavior**
The subscription just stops processing messages.

**Config/Logs/Screenshots**
```
{"@t":"2022-05-10T13:45:28.4862677+02:00","@mt":"Live subscription {subscriptionId} to {streamName} timed out at {streamRevision}; unsubscribing...","@l":"Verbose","@i":4039113944,"@x":"System.OperationCanceledException: The operation was canceled.\r\n   at System.Threading.Channels.AsyncOperation`1.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)\r\n   at EventStore.Core.Services.Transport.Grpc.Enumerators.StreamSubscription.<>c__DisplayClass25_0.<<GoLive>g__OnSubscriptionMessage|1>d.MoveNext() in D:\\a\\TrainStation\\TrainStation\\build\\oss-eventstore\\src\\EventStore.Core\\Services\\Transport\\Grpc\\Enumerators.StreamSubscription.cs:line 351","subscriptionId":"4b77640d-0ee9-411c-a625-4cfefffc9bd2","streamName":"<streamname>","streamRevision":"6652136","SourceContext":"EventStore.Core.Services.Transport.Grpc.Enumerators+StreamSubscription","ProcessId":19448,"ThreadId":38}
```

**EventStore details**
 - EventStore server version: Both 20.10.5 and 21.10.2

 - Operating system: Windows Server 2019

 - EventStore client version (if applicable): EventStore.Client.Grpc.Streams 22.0.0

**Comment by rastrup** [comment]
It does seem that the subscription starts up again but no events are processed, from the normal log file:
```
{"@t":"2022-05-10T13:45:28.4862677+02:00","@mt":"Live subscription {subscriptionId} to {streamName} timed out at {streamRevision}; unsubscribing...","@l":"Verbose","@i":4039113944,"@x":"System.OperationCanceledException: The operation was canceled.\r\n   at System.Threading.Channels.AsyncOperation`1.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)\r\n   at EventStore.Core.Services.Transport.Grpc.Enumerators.StreamSubscription.<>c__DisplayClass25_0.<<GoLive>g__OnSubscriptionMessage|1>d.MoveNext() in D:\\a\\TrainStation\\TrainStation\\build\\oss-eventstore\\src\\EventStore.Core\\Services\\Transport\\Grpc\\Enumerators.StreamSubscription.cs:line 351","subscriptionId":"4b77640d-0ee9-411c-a625-4cfefffc9bd2","streamName":"<streamname>","streamRevision":"6652136","SourceContext":"EventStore.Core.Services.Transport.Grpc.Enumerators+StreamSubscription","ProcessId":19448,"ThreadId":38}
{"@t":"2022-05-10T13:45:28.4862964+02:00","@mt":"Catch-up subscription {subscriptionId} to {streamName}@{streamRevision} running...","@l":"Information","@i":847908818,"subscriptionId":"4b77640d-0ee9-411c-a625-4cfefffc9bd2","streamName":"<streamname>","streamRevision":"6652100","SourceContext":"EventStore.Core.Services.Transport.Grpc.Enumerators+StreamSubscription","ProcessId":19448,"ThreadId":38}
{"@t":"2022-05-10T13:45:28.4863000+02:00","@mt":"Subscription {subscriptionId} to {streamName} reading next page starting from {nextRevision}.","@l":"Verbose","@i":2857299729,"subscriptionId":"4b77640d-0ee9-411c-a625-4cfefffc9bd2","streamName":"<streamname>","nextRevision":"6652100","SourceContext":"EventStore.Core.Services.Transport.Grpc.Enumerators+StreamSubscription","ProcessId":19448,"ThreadId":38}
```

**Comment by rastrup** [comment]
After the subscription timeouts we also start getting hanged calls reading from the stream and have to use a cancellation token to continue.

**Comment by StevenBlair123** [comment]
@rastrup - could you give me the sample code you used for connecting?
I could try it on one of my evenstore databases and see if the same thing happens?
Is it for every stream you see this, or a specific one?

**Comment by rastrup** [comment]
We are connecting with this connection string:
`esdb://<username>:<password>@<servername>:<httpport>&connectionName=<name>&keepAliveInterval=10000&keepAliveTimeout=10000&tls=true&tlsVerifyCert=false`

What we are seeing is this:
We have a test that test our system under some load.
It starts up our system which is a Web Server with 15 background services each containing 1 Subscription to a stream or a projection.
Then we start adding load and then after 5-10 minutes the subscriptions to the streams with the most load start to time out as shown in the log. The log says the subscription is restarted (see above) but no more events come through. What is more the stream/projections starts to cause other reads to hang (we are using cancellation tokens with timeout to move forward).
This is found during migration to the Grpc connection from the Tcp connection. The subscriptions run fine using a Tcp connection.

**Comment by rastrup** [comment]
> @rastrup - I have created a test harness, which you cna find here: https://github.com/StevenBlair123/evenstore_examples I tried running 100 catchup subscriptions, included various timeouts. CPU usage was maxed out and I could not get this problem replicated.
> 
> Could you try using my harness on your database and see if you can replicate it? Try it without any additional writing being done, then enable writing. If it looks like the writing of events has some link to the problem, I cna look at updating this harness to write lots of data at the same time.
> 
> Let me know how you get on.
> 
> Early in my testing, I thought I had replicated the issue. A few times in a row, my process stopped after about 5 minutes (15 catchups). However, what I hadn't realised was the catchups had actually finished!

I am not sure what you mean by using your harness. Do you want me to use it while I run my load test?

The timeouts are not user generated. I don't think we have any part of our event handling that takes very long time. And as written before these timeouts do not occur over a Tcp connection.
We also only see the timeouts on the server side.

**Comment by StevenBlair123** [comment]
@rastrup - yeah if you could my harness against your database. I am keen to see if the same problem occurs (it might make it easier to diagnose)
If you think the timeouts are not needed, its very easy to remove on the harness.

Ultimately, what I would like is you being able to replicate it on the harness, and I cna try to debug it on an database I have (I have the source code open so can debug it)

**Comment by rastrup** [comment]
I have tried the harness against the database without anything else running on it and it just reads all the events.
I have also tried the harness against the database while our solution is in this timeout state and it just hangs and reads no events.

**Comment by StevenBlair123** [comment]
Any specific errors (server / client)
I wonder if the next route could be to enhance the harness to write lots of events while it's running.

What do you think?

**Comment by StevenBlair123** [comment]
@rastrup - I added a concurrent writer into the harness.
Could you try this? 
Obviously not against any production database, unless you want to truncate the test events.

I have ramped up the amount of traffic hitting my local database to 25 catchups and 50 concurrent writers.
Plenty of "SLOW QUEUE MSG" but all catchups behaving as expected.

**Comment by rastrup** [comment]
> @rastrup - I added a concurrent writer into the harness. Could you try this? Obviously not against any production database, unless you want to truncate the test events.
> 
> I have ramped up the amount of traffic hitting my local database to 25 catchups and 50 concurrent writers. Plenty of "SLOW QUEUE MSG" but all catchups behaving as expected.

I will see what I can do, since the issue has been discovered in a test scenario we can do a lot of things, including running custom coded event store.

**Comment by AMontagu** [comment]
I randomly arrive at this issue but it seem to be a lot like mine if you are using grpc client: https://github.com/grpc/grpc/issues/29561

How many time your tiemout appear ? (me it's 2-3 days without message received) Are you on kubernetes ? Istio ?

**Comment by rastrup** [comment]
> I randomly arrive at this issue but it seem to be a lot like mine if you are using grpc client: [grpc/grpc#29561](https://github.com/grpc/grpc/issues/29561)
> 
> How many time your tiemout appear ? (me it's 2-3 days without message received) Are you on kubernetes ? Istio ?

I am on an internal server at a company. The issue happens during a load and integration test. We are for this reason not using Grpc in subscriptions in production.

**Comment by AMontagu** [comment]
So it's not the same issue than me. Sorry for the step in your issue.

**Comment by rastrup** [comment]
> @rastrup - I added a concurrent writer into the harness. Could you try this? Obviously not against any production database, unless you want to truncate the test events.
> 
> I have ramped up the amount of traffic hitting my local database to 25 catchups and 50 concurrent writers. Plenty of "SLOW QUEUE MSG" but all catchups behaving as expected.

I have tried the updated harness, still the same. It runs fine when running on it's own, but once our eventstore is in the faulty state it halts.

Something to note: it seems from the error log that the timeout occurs inside the Event Store and not in user code, is that a correct observation?
```
System.OperationCanceledException: The operation was canceled.
   at System.Threading.Channels.AsyncOperation`1.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at EventStore.Core.Services.Transport.Grpc.Enumerators.StreamSubscription.<>c__DisplayClass25_0.<<GoLive>g__OnSubscriptionMessage|1>d.MoveNext() in D:\\a\\TrainStation\\TrainStation\\build\\oss-eventstore\\src\\EventStore.Core\\Services\\Transport\\Grpc\\Enumerators.StreamSubscription.cs:line 351
```

**Comment by StevenBlair123** [comment]
How big is your database when it get's into this faulted state?
Is it something that could be shared?

**Comment by rastrup** [comment]
> How big is your database when it get's into this faulted state? Is it something that could be shared?

The Event Store in the test scenario is around 66 GB. It cannot be shared from the client but we can run a custom version of the Event Store with added debug info and others (branch 20.10.5 preferred) and probably also do a shared session.

**Comment by rastrup** [comment]
I am looking inside Enumerators.StreamSubscription.cs in the EventStore Repo and from the error message it is 
`await liveEvents.Writer.WriteAsync(appeared.Event, cts.Token).ConfigureAwait(false);`
on line 363 that times out after the 1 second timeout.
What does this line do and what could make it timeout?

**Comment by StevenBlair123** [comment]
@rastrup 

I feel like I can replicate your problem now (I have the database running from source)
My gut feeling is, the code in the catch block and CancellationToken are not related to the problem.

My guess is, the call to Unsubscribe cleans up the server side, but does not notify the client (and I cannot work out how that actually happens)

The reason I beleive this, is I modified the code to do the following:

Once we have been notified that we caught up (ClientMessage.SubscriptionConfirmation) , I force the a call to Unsubscribe.
This produces the message "dropped: Unsubscribed" but the client is not notified.
If I post another event onto the stream I was susbcribed to, my catchup client does not receive anything.

That suggests thtat the Catchup subscription has been dropped, but the client did not get informed.

Would you agree this replicates your problem?

**Comment by rastrup** [comment]
> @rastrup
> 
> I feel like I can replicate your problem now (I have the database running from source) My gut feeling is, the code in the catch block and CancellationToken are not related to the problem.
> 
> My guess is, the call to Unsubscribe cleans up the server side, but does not notify the client (and I cannot work out how that actually happens)
> 
> The reason I beleive this, is I modified the code to do the following:
> 
> Once we have been notified that we caught up (ClientMessage.SubscriptionConfirmation) , I force the a call to Unsubscribe. This produces the message "dropped: Unsubscribed" but the client is not notified. If I post another event onto the stream I was susbcribed to, my catchup client does not receive anything.
> 
> That suggests thtat the Catchup subscription has been dropped, but the client did not get informed.
> 
> Would you agree this replicates your problem?

That sounds about right. Also the stream locks up for other readers.

**Comment by timothycoleman** [comment]
Hi @rastrup  @StevenBlair123, i think quite likely https://github.com/EventStore/EventStore/pull/3532 will fix the behaviour you've seen here

**Comment by ylorph** [comment]
@rastrup , @StevenBlair123 : the fix is available in 21.10.7 (LTS release)  & 22.6 (interim release)

**Comment by rastrup** [comment]
> @rastrup , @StevenBlair123 : the fix is available in 21.10.7 (LTS release) & 22.6 (interim release)

Confirmed, 21.10.7 fixes this issue.
